### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by MarekMacko**
- removed unnecessary properties in response models,
- separate models for ui,
- modules reorganization,
- jacoco test report,
- unit tests with <a href="https://github.com/fabioCollini/DaggerMock">DaggerMock</a> and <a href="https://github.com/robolectric/robolectric">Robolectric</a>

### CODE REVIEW

File: app/src/main/java/com/marekmacko/kotlinapp/ForecastListAdapter.kt [hunk 1]
```diff
@@ -23,41 +20,25 @@ class ForecastListAdapter(private val itemClick: (DailyForecast) -> Unit)
...
 
     override fun onBindViewHolder(holder: ViewHolder, position: Int) =
-            holder.bindDailyForecast(weeklyForecast!![position])
+            holder.bindForecast(weeklyForecast!![position])
 
     override fun getItemCount(): Int = weeklyForecast?.size ?: 0
 
-    fun setItems(weeklyForecast: WeeklyForecast) {
+    fun setWeeklyForecast(weeklyForecast: List<Forecast>) {
         this.weeklyForecast = weeklyForecast
```
Comment (sswierczek) [inline app/src/main/java/com/marekmacko/kotlinapp/ForecastListAdapter.kt]: BTW it is better to keep the same reference to list in adapter, clear old one, and addAll new elements. When changing reference notify methods could work not properly.

File: app/src/main/java/com/marekmacko/kotlinapp/ForecastListAdapter.kt [hunk 2]
```diff
@@ -23,41 +20,25 @@ class ForecastListAdapter(private val itemClick: (DailyForecast) -> Unit)
     }
 
     override fun onBindViewHolder(holder: ViewHolder, position: Int) =
-            holder.bindDailyForecast(weeklyForecast!![position])
+            holder.bindForecast(weeklyForecast!![position])
```
Comment (sswierczek) [inline app/src/main/java/com/marekmacko/kotlinapp/ForecastListAdapter.kt]: Can weekly forecast be non-null type?

File: app/src/main/java/com/marekmacko/kotlinapp/data/response/Weather.kt [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package com.marekmacko.kotlinapp.data.response
+
+import com.google.gson.annotations.SerializedName
+
+data class Weather(val description: String,
+                   @SerializedName("icon") private val iconCode: String) {
```
Comment (sswierczek) [inline app/src/main/java/com/marekmacko/kotlinapp/data/response/Weather.kt]: You could just name it "icon"

File: app/src/main/java/com/marekmacko/kotlinapp/repository/WeatherRepository.kt [hunk 4]
```diff
@@ -1,11 +1,20 @@
...
-    fun getWeeklyForecast(onLoad: (WeeklyForecast) -> Unit, onError: (String) -> Unit): Disposable =
-            remoteWeatherRepository.getWeeklyForecast(onLoad, onError)
+    companion object {
+        private const val API_ZIP_CODE = "94043"
+    }
+
+    fun getWeeklyForecast(): Observable<List<Forecast>> =
+            weatherService.getWeeklyForecast(API_ZIP_CODE)
+                    .map {
+                        ModelMapper.convertResponseToForecast(it)
```
Comment (sswierczek) [inline app/src/main/java/com/marekmacko/kotlinapp/repository/WeatherRepository.kt]: Static methods are hard to mock. You could just create class method and inject instance here

File: app/src/main/java/com/marekmacko/kotlinapp/util/Utils.kt [hunk 5]
```diff
@@ -10,7 +10,7 @@ class Utils {
         fun isNetworkAvailable(context: Context): Boolean {
             val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
             val networkInfo = connectivityManager.activeNetworkInfo
-            return networkInfo?.isConnected ?: false
+            return networkInfo?.isConnected == true
```
Comment (sswierczek) [inline app/src/main/java/com/marekmacko/kotlinapp/util/Utils.kt:13]: Couldn't be just return?

File: app/src/main/res/layout/fragment_weekly_forecast.xml [hunk 6]
```diff
@@ -26,4 +28,29 @@
...
+        android:visibility="invisible"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintLeft_toLeftOf="parent"
+        app:layout_constraintRight_toRightOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        tools:visibility="visible" />
+
+    <ImageView
+        android:id="@+id/refreshIconView"
+        android:layout_width="48dp"
```
Comment (sswierczek) [inline app/src/main/res/layout/fragment_weekly_forecast.xml]: dimens.xml?

File: app/src/test/java/com/marekmacko/kotlinapp/api/WeatherServiceTest.kt [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import io.reactivex.Observable
+import io.reactivex.observers.TestObserver
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.Mockito
+
+@RunWith(DataProviderRunner::class)
+class WeatherServiceTest {
+
+    private val weatherService: WeatherService = mock()
```
Comment (sswierczek) [inline app/src/test/java/com/marekmacko/kotlinapp/api/WeatherServiceTest.kt]: You shouldn't mock object you tested :)

File: app/src/test/java/com/marekmacko/kotlinapp/data/WeatherTest.kt [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+import com.marekmacko.kotlinapp.data.response.Weather
+import org.junit.Test
+
+import org.junit.Assert.*
+
+class WeatherTest {
+
+    @Test
+    fun getIconUrl() {
+        val weather = Weather(MockData.DESCRIPTION, MockData.ICON_CODE)
```
Comment (sswierczek) [inline app/src/test/java/com/marekmacko/kotlinapp/data/WeatherTest.kt]: It is better to keep 3 parts (or 2) in tests GIVEN WHEN THEN. Think you should test also some invalid data like empty and/or null input values. 

This comment is valid for all tests :)

File: app/src/test/java/com/marekmacko/kotlinapp/mvp/WeatherPresenterTest.kt [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    lateinit var weatherPresenter: WeatherPresenter
+
+    @Test
+    @UseDataProvider("getWeeklyForecastArray", location = arrayOf(MockData::class))
+    fun showLoadingIsCalledOnFetchForecast(forecastList: List<Forecast>) {
+        whenever(weatherRepository.getWeeklyForecast()).thenReturn(Observable.just(forecastList))
+
+        weatherPresenter.fetchForecast()
+
+        verify(view, times(1)).showLoading()
```
Comment (sswierczek) [inline app/src/test/java/com/marekmacko/kotlinapp/mvp/WeatherPresenterTest.kt]: times(1) can be omitted

File: app/src/test/java/com/marekmacko/kotlinapp/ui/DailyForecastFragmentTest.kt [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+                .commit()
+        activityController.create().start().resume().visible()
+    }
+
+    @Test
+    fun titleIsSetup() = assertEquals(MockData.DATE_STRING, fragment.activity.title)
+
+    @Test
+    fun forecastDataIsSetup() {
+        assertEquals(MockData.DESCRIPTION, fragment.descriptionView.text)
```
Comment (sswierczek) [inline app/src/test/java/com/marekmacko/kotlinapp/ui/DailyForecastFragmentTest.kt]: I would recommend separate tests for these it is hard to follow what should be done with many asserts

**Review by sswierczek — APPROVED** [review]

File: app/src/main/java/com/marekmacko/kotlinapp/util/Utils.kt (hunk 5 above)
Comment (MarekMacko) [reply app/src/main/java/com/marekmacko/kotlinapp/util/Utils.kt:13]: No, because there could be situation when `networkInfo` will be null and this will be inconsistent with return type.

File: app/src/main/res/layout/fragment_weekly_forecast.xml (hunk 6 above)
Comment (MarekMacko) [reply app/src/main/res/layout/fragment_weekly_forecast.xml]: I omitted this intentionally, because I want to focus only on kotlin 😃
